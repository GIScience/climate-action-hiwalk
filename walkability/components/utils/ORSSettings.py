import openrouteservice
from dataclasses import dataclass


@dataclass
class ORSSettings:
    # For future reference maybe check this suggestion: https://gitlab.heigit.org/climate-action/plugins/walkability/-/merge_requests/82#note_61406
    client: openrouteservice.Client

    snapping_rate_limit: int = 100
    snapping_request_size_limit: int = 4999

    directions_rate_limit: int = 40
    directions_waypoint_limit: int = 50
