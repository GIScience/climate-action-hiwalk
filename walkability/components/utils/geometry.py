import geopandas as gpd
import shapely
from pyproj import CRS, Transformer
from shapely.ops import transform

WGS84 = CRS('EPSG:4326')


def get_buffered_aoi(aoi: shapely.MultiPolygon, distance: float) -> shapely.MultiPolygon:
    utm = get_utm_zone(aoi)

    geographic_projection_function = Transformer.from_crs(WGS84, utm, always_xy=True).transform
    wgs84_projection_function = Transformer.from_crs(utm, WGS84, always_xy=True).transform
    projected_aoi = transform(geographic_projection_function, aoi)
    buffered_aoi: shapely.MultiPolygon = projected_aoi.buffer(distance)
    return transform(wgs84_projection_function, buffered_aoi)


def get_utm_zone(aoi: shapely.MultiPolygon) -> CRS:
    return gpd.GeoSeries(data=aoi, crs=WGS84).estimate_utm_crs()
