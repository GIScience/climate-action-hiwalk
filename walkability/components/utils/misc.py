import logging
from enum import Enum
from typing import Dict, Tuple, List, Optional

import geopandas as gpd
import matplotlib as mpl
import pandas as pd
import shapely
from numpy import number
from ohsome import OhsomeClient
from pydantic_extra_types.color import Color

log = logging.getLogger(__name__)


class PathCategory(Enum):
    DESIGNATED = 'Designated'
    DESIGNATED_SHARED_WITH_BIKES = 'Shared with bikes'
    SHARED_WITH_MOTORIZED_TRAFFIC_LOW_SPEED = 'Shared with slow cars'
    SHARED_WITH_MOTORIZED_TRAFFIC_MEDIUM_SPEED = 'Shared with medium speed cars'
    SHARED_WITH_MOTORIZED_TRAFFIC_HIGH_SPEED = 'Shared with fast cars'
    SHARED_WITH_MOTORIZED_TRAFFIC_UNKNOWN_SPEED = 'Shared with cars of unknown speed'
    NOT_WALKABLE = 'Not walkable'
    UNKNOWN = 'Unknown'


PATH_RATING_MAP = {
    PathCategory.DESIGNATED: 1.0,
    PathCategory.DESIGNATED_SHARED_WITH_BIKES: 0.8,
    PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_LOW_SPEED: 0.6,
    PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_MEDIUM_SPEED: 0.4,
    PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_HIGH_SPEED: 0.2,
    PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_UNKNOWN_SPEED: 0.1,
    PathCategory.NOT_WALKABLE: 0.0,
    PathCategory.UNKNOWN: None,
}


class PavementQuality(Enum):
    GOOD = 'good'
    POTENTIALLY_GOOD = 'potentially_good'
    MEDIOCRE = 'mediocre'
    POTENTIALLY_MEDIOCRE = 'potentially_mediocre'
    POOR = 'poor'
    UNKNOWN = 'unknown'


PAVEMENT_QUALITY_RATING_MAP = {
    PavementQuality.GOOD: 1.0,
    PavementQuality.POTENTIALLY_GOOD: 0.8,
    PavementQuality.MEDIOCRE: 0.5,
    PavementQuality.POTENTIALLY_MEDIOCRE: 0.3,
    PavementQuality.POOR: 0.0,
    PavementQuality.UNKNOWN: None,
}


class SmoothnessCategory(Enum):
    GOOD = 'good'
    MEDIOCRE = 'mediocre'
    POOR = 'poor'
    VERY_POOR = 'very_poor'
    UNKNOWN = 'unknown'


SMOOTHNESS_CATEGORY_RATING_MAP = {
    SmoothnessCategory.GOOD: 1.0,
    SmoothnessCategory.MEDIOCRE: 0.5,
    SmoothnessCategory.POOR: 0.2,
    SmoothnessCategory.VERY_POOR: 0.0,
    SmoothnessCategory.UNKNOWN: None,
}


class SurfaceType(Enum):
    CONTINUOUS_PAVEMENT = 'continuous_pavement'
    MODULAR_PAVEMENT = 'modular_pavement'
    COBBLESTONE = 'cobblestone'
    OTHER_PAVED = 'other_paved_surfaces'
    GRAVEL = 'gravel'
    GROUND = 'ground'
    OTHER_UNPAVED = 'other_unpaved_surfaces'
    UNKNOWN = 'unknown'


SURFACE_TYPE_RATING_MAP = {
    SurfaceType.CONTINUOUS_PAVEMENT: 1.0,
    SurfaceType.MODULAR_PAVEMENT: 0.85,
    SurfaceType.COBBLESTONE: 0.6,
    SurfaceType.OTHER_PAVED: 0.5,
    SurfaceType.GRAVEL: 0.4,
    SurfaceType.GROUND: 0.2,
    SurfaceType.OTHER_UNPAVED: 0.0,
    SurfaceType.UNKNOWN: None,
}


def fetch_osm_data(aoi: shapely.MultiPolygon, osm_filter: str, ohsome: OhsomeClient) -> gpd.GeoDataFrame:
    elements = ohsome.elements.geometry.post(
        bpolys=aoi, clipGeometry=True, properties='tags', filter=osm_filter
    ).as_dataframe()
    elements = elements.reset_index(drop=False)
    return elements[['@osmId', 'geometry', '@other_tags']]


def _dict_to_legend(d: dict, cmap_name: str = 'coolwarm_r') -> Dict[str, Color]:
    data = pd.DataFrame.from_records(data=list(d.items()), columns=['category', 'rating'])
    data['color'] = generate_colors(color_by=data.rating, cmap_name=cmap_name, min=0.0, max=1.0)
    data['category'] = data.category.apply(lambda cat: cat.value)
    return dict(zip(data['category'], data['color']))


def get_path_rating_legend() -> Dict[str, Color]:
    return _dict_to_legend(PATH_RATING_MAP, cmap_name='coolwarm_r')


def get_surface_quality_legend() -> Dict[str, Color]:
    return _dict_to_legend(PAVEMENT_QUALITY_RATING_MAP, cmap_name='coolwarm_r')


def get_smoothness_legend() -> Dict[str, Color]:
    return _dict_to_legend(SMOOTHNESS_CATEGORY_RATING_MAP, cmap_name='coolwarm_r')


def get_surface_type_legend() -> Dict[str, Color]:
    return _dict_to_legend(SURFACE_TYPE_RATING_MAP, cmap_name='tab10')


def generate_colors(
    color_by: pd.Series, cmap_name: str, min: number | None = None, max: number | None = None
) -> list[Color]:
    """
    Function to generate a list of colors based on a linear normalization for each element in `color_by`.
    ## Params
    :param:`color_by`: `pandas.Series` with numerical values to map colors to.
    :param:`cmap_name`: name of matplotlib colormap approved in climatoology.
    :param:`min`: optional minimal value of the normalization, numerical or `None`. If `None` minimum value of `color_by` is used.
    :param:`max`: optional maxmial value of the normalization, numerical or `None`. If `None` maximum value of `color_by` is used.
    ## Returns
    :return:`mapped_colors`: list of colors matching values of `color_by`
    """
    color_by = color_by.astype(float)
    if min is None:
        min = color_by.min()
    if max is None:
        max = color_by.max()

    norm = mpl.colors.Normalize(vmin=min, vmax=max)
    cmap = mpl.colormaps[cmap_name]
    cmap.set_bad(color='#808080')

    mapped_colors = [Color(mpl.colors.to_hex(cmap(norm(val)))) for val in color_by]
    return mapped_colors


def get_first_match(ordered_keys: List[str], tags: Dict[str, str]) -> Tuple[Optional[str], Optional[str]]:
    match_key = None
    match_value = None
    for key in ordered_keys:
        match_value = tags.get(key)
        if match_value:
            match_key = key
            break

    return match_key, match_value


def ohsome_filter(geometry_type: str) -> str:
    if geometry_type == 'relation':
        # currently unused, here as a blueprint if we want to query for relations
        return str(f'type:{geometry_type} and route=ferry')

    return str(
        f'geometry:{geometry_type} and '
        '(highway=* or route=ferry or railway=platform or '
        '(waterway=lock_gate and foot in (yes, designated, official, permissive))) and not '
        '(footway=separate or sidewalk=separate or sidewalk:both=separate or '
        '(sidewalk:right=separate and sidewalk:left=separate) or '
        '(sidewalk:right=separate and sidewalk:left=no) or (sidewalk:right=no and sidewalk:left=separate))'
    )


def safe_string_to_float(potential_number: str | float) -> float:
    try:
        return float(potential_number)
    except (TypeError, ValueError):
        return -1
