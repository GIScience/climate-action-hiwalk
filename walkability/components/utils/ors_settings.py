from dataclasses import dataclass

import openrouteservice


@dataclass
class ORSSettings:
    # For future reference maybe check this suggestion: https://gitlab.heigit.org/climate-action/plugins/walkability/-/merge_requests/82#note_61406
    client: openrouteservice.Client

    snapping_rate_limit: int = 100
    snapping_request_size_limit: int = 4999

    directions_rate_limit: int = 40
    directions_waypoint_limit: int = 50

    coordinate_precision: float = 0.000001
    isochrone_max_batch_size: int = 5
