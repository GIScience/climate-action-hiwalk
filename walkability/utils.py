from enum import Enum
from typing import Dict, Union, Callable, Any, Tuple
from urllib.parse import parse_qsl

import geopandas as gpd
import matplotlib
import pandas as pd
import shapely
from matplotlib.colors import to_hex
from ohsome import OhsomeClient
from pydantic import BaseModel
from pydantic_extra_types.color import Color
from requests import PreparedRequest
from shapely import LineString


class FilterGroups(BaseModel):
    exclusive: str
    explicit: str
    probable_yes: str
    probable_no: str
    inaccessible: str


class Rating(Enum):
    EXCLUSIVE = 1.0
    EXPLICIT = 0.75
    PROBABLE_YES = 0.5
    PROBABLE_NO = 0.25
    INACCESSIBLE = 0.0


def construct_filters() -> Dict[Rating, str]:
    # TODO: Remove whitespace before sending to ohsome API
    # Potential: potentially walkable features (to be restricted by AND queries)
    potential_highway_values = (
        'primary',
        'primary_link',
        'secondary',
        'secondary_link',
        'tertiary',
        'tertiary_link',
        'unclassified',
        'residential',
        'service',
        'track',
        'road',
        'cycleway',
        'platform',
    )
    potential = f"""
    highway in ({','.join(potential_highway_values)}) or
    route=ferry
    """
    # For documentation:
    # ignored_primary = 'highway in (motorway,trunk,motorway_link,trunk_link,
    # primary_link,secondary_link,tertiary_link,bus_guideway,escape,raceway,busway,
    # brideleway,via_ferrata,cicleway'
    # ignored_secondary = 'sidewalk=no'
    ignore = """
    footway in (separate,no) or
    sidewalk=separate or
    sidewalk:left=separate or
    sidewalk:right=separate or
    sidewalk:both=separate or
    access in (no,private,permit,military,delivery,customers) or
    foot in (no,private,use_sidepath,discouraged,destination)
    """

    # Exclusive: Only for pedestrians
    # secondary tags
    exclusive_secondary = """
    (
        foot in (designated,official) or
        foot!=*
    ) or
    (
        footway in (access_aisle,alley,residential,link,path) or
        footway!=*
    )
    """

    _exclusive = f"""
    highway in (pedestrian,steps,corridor) or
    (
        highway=footway and
        ({exclusive_secondary})
    ) or
    (
        highway=path and
        ({exclusive_secondary})
    )
    """

    exclusive = f"""
    ({_exclusive}) and not
    ({ignore})
    """

    # Explicit: For pedestrians but not only (E.g. shared with bicycle)
    explicit_foot = 'foot in (yes,permissive,designated,official)'
    # secondary tags
    explicit_secondary = f"""
    {explicit_foot} or
    footway in (sidewalk,crossing,traffic_island,yes)
    """

    _explicit = f"""
    railway=platform or
    highway=living_street or
    (
        highway=footway and
        ({explicit_secondary})
    ) or
    (
        highway=path and
        ({explicit_secondary})
    ) or
    (
        ({potential}) and
        (
            sidewalk in (both,left,right,yes,lane) or
            sidewalk:left=yes or
            sidewalk:right=yes or
            sidewalk:both=yes or
            ({explicit_foot})
        )
    )
    """
    explicit = f"""
    ({_explicit}) and not
    (
        ({ignore}) or
        ({_exclusive})
    )
    """

    _inaccessible = f"""
    ({potential}) and
    (
        sidewalk=no or
        sidewalk:both=no or
        (sidewalk:left=no and sidewalk:right=no) or
        sidewalk=none or
        sidewalk:both=none or
        (sidewalk:left=none and sidewalk:right=none)
    )
    """
    inaccessible = f"""
    ({_inaccessible}) and not
    (
        ({ignore}) or
        ({_exclusive}) or
        ({_explicit})
    )
    """

    # Probable_yes: (E.g. forest tracks)
    _probable_yes = """
    highway in (track,service,path) or
    man_made=pier
    """
    probable_yes = f"""
    ({_probable_yes}) and not
    (
        ({ignore}) or
        ({_exclusive}) or
        ({_explicit}) or
        ({_inaccessible})
    )
    """

    probable_no = f"""
    ({potential}) and not
    (
        ({ignore}) or
        ({_exclusive}) or
        ({_explicit}) or
        ({_inaccessible}) or
        ({_probable_yes})
    )
    """
    # Remove empty lines for better readability
    exclusive = ''.join([s for s in exclusive.strip().splitlines(keepends=True) if s.strip()])
    explicit = ''.join([s for s in explicit.strip().splitlines(keepends=True) if s.strip()])
    probable_yes = ''.join([s for s in probable_yes.strip().splitlines(keepends=True) if s.strip()])
    probable_no = ''.join([s for s in probable_no.strip().splitlines(keepends=True) if s.strip()])
    inaccessible = ''.join([s for s in inaccessible.strip().splitlines(keepends=True) if s.strip()])
    return {
        Rating.EXCLUSIVE: exclusive,
        Rating.EXPLICIT: explicit,
        Rating.PROBABLE_YES: probable_yes,
        Rating.PROBABLE_NO: probable_no,
        Rating.INACCESSIBLE: inaccessible,
    }


def fetch_osm_data(
    aoi: shapely.MultiPolygon, osm_filter: str, rating: Rating, ohsome: OhsomeClient
) -> gpd.GeoDataFrame:
    elements = ohsome.elements.geometry.post(
        bpolys=aoi, clipGeometry=True, properties=None, filter=osm_filter
    ).as_dataframe()
    elements = elements.reset_index(drop=True)
    elements['category'] = rating
    return elements[['category', 'geometry']]


def boost_route_members(aoi: shapely.MultiPolygon, paths_line: gpd.GeoDataFrame, ohsome: OhsomeClient) -> pd.Series:
    trails = fetch_osm_data(aoi, 'route in (foot,hiking)', Rating.EXPLICIT, ohsome)

    trails.geometry = trails.geometry.apply(lambda geom: fix_geometry_collection(geom))

    paths_line = paths_line.copy()
    paths_line = gpd.sjoin(
        paths_line,
        trails,
        lsuffix='path',
        rsuffix='trail',
        how='left',
        predicate='within',
    )
    paths_line = paths_line[~paths_line.index.duplicated(keep='first')]
    paths_line.loc[paths_line.category_trail.isnull(), 'category_trail'] = Rating.INACCESSIBLE

    return paths_line.apply(
        lambda row: row.category_path if row.category_path.value >= row.category_trail.value else row.category_trail,
        axis=1,
    )


def fix_geometry_collection(
    geom: shapely.Geometry,
) -> Union[shapely.LineString, shapely.MultiLineString]:
    # Hack due to https://github.com/GIScience/oshdb/issues/463
    if geom.geom_type == 'GeometryCollection':
        inner_geoms = []
        for inner_geom in geom.geoms:
            if inner_geom.geom_type in ('LineString', 'MultiLineString'):
                inner_geoms.append(inner_geom)
        geom = shapely.union_all(inner_geoms)

    if geom.geom_type in ('LineString', 'MultiLineString'):
        return geom
    else:
        return LineString()


def get_color(categories: pd.Series) -> pd.Series:
    cmap = matplotlib.colormaps.get_cmap('RdYlGn')
    return categories.apply(lambda c: Color(to_hex(cmap(c.value))))


def filter_start_matcher(filter_start: str) -> Callable[..., Any]:
    def match(request: PreparedRequest) -> Tuple[bool, str]:
        request_body = request.body
        qsl_body = dict(parse_qsl(request_body, keep_blank_values=False)) if request_body else {}

        if request_body is None:
            return False, 'The given request has no body'
        elif qsl_body.get('filter') is None:
            return False, 'Filter parameter not set'
        else:
            valid = qsl_body.get('filter', '').startswith(filter_start)
            return (True, '') if valid else (False, f'The filter parameter does not start with {filter_start}')

    return match
