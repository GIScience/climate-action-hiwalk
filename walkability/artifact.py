from typing import List, Dict

import geopandas as gpd
import pandas as pd
import shapely
from climatoology.base.artifact import (
    _Artifact,
    create_geojson_artifact,
    create_chart_artifact,
    Chart2dData,
    ContinuousLegendData,
)
from climatoology.base.computation import ComputationResources

from walkability.input import PathRating
from walkability.utils import (
    get_color,
    get_single_color,
    PavementQualityRating,
    generate_detailed_pavement_quality_mapping_info,
)


def build_paths_artifact(
    paths_line: gpd.GeoDataFrame,
    paths_polygon: gpd.GeoDataFrame,
    ratings: PathRating,
    clip_aoi: shapely.MultiPolygon,
    resources: ComputationResources,
    cmap_name: str = 'RdYlGn',
) -> _Artifact:
    paths_line = paths_line.clip(clip_aoi, keep_geom_type=True)
    paths_polygon = paths_polygon.clip(clip_aoi, keep_geom_type=True)
    sidewalks = pd.concat([paths_line, paths_polygon], ignore_index=True)

    sidewalks['color'] = get_color(sidewalks.rating, cmap_name)
    return create_geojson_artifact(
        features=sidewalks.geometry,
        layer_name='Walkable',
        caption='The layer displays paths in four categories: '
        'a) paths dedicated to pedestrians exclusively '
        'b) paths that are explicitly meant for pedestrians but may be shared with other traffic (e.g. a '
        'road with a sidewalk) '
        'c) paths that probably are walkable but the true status is unknown (e.g. a dirt road) '
        'd) paths that are not walkable but could be (e.g. a residential road without sidewalk).',
        description='The layer excludes paths that are not walkable by definition such as motorways or cycle ways. '
        'The data source is OpenStreetMap.',
        label=sidewalks.category.apply(lambda r: r.name).to_list(),
        color=sidewalks.color.to_list(),
        legend_data={rating[0]: get_single_color(rating[1]) for rating in ratings},
        resources=resources,
        filename='walkable',
    )


def build_connectivity_artifact(
    connectivity: gpd.GeoDataFrame,
    clip_aoi: shapely.MultiPolygon,
    resources: ComputationResources,
    cmap_name: str = 'RdYlGn',
) -> _Artifact:
    connectivity = connectivity.clip(clip_aoi, keep_geom_type=True)
    color = get_color(connectivity.connectivity, cmap_name).to_list()
    legend = ContinuousLegendData(
        cmap_name=cmap_name, ticks={'Low Connectivity': 0, 'Medium Connectivity': 0.5, 'High Connectivity': 1}
    )

    return create_geojson_artifact(
        features=connectivity.geometry,
        layer_name='Connectivity',
        primary=False,
        filename='connectivity',
        caption='Map of connectivity scores.',
        description='Each path is evaluated based on the reachability of other paths within the area of interest. '
        'Reachability or connectivity is defined as the share of locations that can be reached by foot in '
        'reference to an optimum where all locations are directly connected "as the crow flies".',
        label=connectivity.connectivity.to_list(),
        color=color,
        legend_data=legend,
        resources=resources,
    )


def build_pavement_quality_artifact(
    paths_line: gpd.GeoDataFrame,
    resources: ComputationResources,
    cmap_name: str = 'RdYlGn',
) -> _Artifact:
    paths_line['color'] = get_color(paths_line.quality.map(PavementQualityRating), cmap_name)
    return create_geojson_artifact(
        features=paths_line.geometry,
        layer_name='Pavement Quality',
        caption='The layer displays the pavement quality for the accessible paths of the walkable layer.',
        description='Based on the values of the `smoothness`, `surface` and `tracktype` tags of OpenStreetMap (in order of importance). '
        'If there is no specification for the pavement of non-exlusive footways, the quality of accompanying roads is adopted if available and labelled as *potential*. '
        'Some surface types, such as gravel, are also labelled *potential* as they can exhibit a wide variation in their maintenance status (see table below).\n\n'
        'Full list of tag-value-ranking combinations:\n\n'
        '' + generate_detailed_pavement_quality_mapping_info(),
        label=paths_line.quality.apply(lambda r: r.name).to_list(),
        color=paths_line.color.to_list(),
        legend_data={rating: get_single_color(PavementQualityRating[rating]) for rating in PavementQualityRating},
        resources=resources,
        filename='pavement_quality',
    )


def build_areal_summary_artifacts(
    regional_aggregates: Dict[str, Chart2dData], resources: ComputationResources
) -> List[_Artifact]:
    chart_artifacts = []
    for region, data in regional_aggregates.items():
        chart_artifact = create_chart_artifact(
            data=data,
            title=region,
            caption='The distribution of paths categories for this administrative area. '
            f'The total length of paths in this area is {sum(data.y)}km',
            resources=resources,
            filename=f'aggregation_{region}',
            primary=False,
        )
        chart_artifacts.append(chart_artifact)
    return chart_artifacts
