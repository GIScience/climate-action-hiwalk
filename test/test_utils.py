from functools import partial
import logging
import urllib.parse

import geopandas as gpd
import geopandas.testing
import pandas as pd
import pytest
import shapely
from approvaltests.approvals import verify
from ohsome import OhsomeClient
from pydantic_extra_types.color import Color
from shapely.testing import assert_geometries_equal
from urllib3 import Retry

from walkability.utils import (
    boost_route_members,
    construct_filters,
    PathCategory,
    fetch_osm_data,
    fix_geometry_collection,
    get_color,
    generate_detailed_pavement_quality_mapping_info,
    apply_path_category_filters,
)


validation_objects = {
    PathCategory.DEDICATED_EXCLUSIVE: {'way/84908668', 'way/243233105', 'way/27797959'},
    # https://www.openstreetmap.org/way/98453212 foot=designated
    # https://www.openstreetmap.org/way/84908668 highway=pedestrian
    # https://www.openstreetmap.org/way/27797959 railway=platform
    PathCategory.DEDICATED_SEPARATED: {'way/98453212', 'way/184725322', 'way/118975501'},
    # https://www.openstreetmap.org/way/184725322 sidewalk:right=right and sidewalk:left=separate
    # https://www.openstreetmap.org/way/243233105 highway=footway,
    # https://www.openstreetmap.org/way/118975501 foot=designated and bicycle=designated and segregated=yes
    PathCategory.SHARED_WITH_BIKES: {'way/25806383', 'way/148612595', 'way/715905259'},
    # https://www.openstreetmap.org/way/25806383 bicycle=designated & foot=designated
    # https://www.openstreetmap.org/way/148612595/history/16 bicycle=yes & highway=residential
    # https://www.openstreetmap.org/way/715905259 highway=track
    PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_LOW_SPEED: {'way/25149880', 'way/14193661'},
    # https://www.openstreetmap.org/way/25149880 highway=service
    # https://www.openstreetmap.org/way/14193661 highway=living_street
    PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_MEDIUM_SPEED: {'way/28890081', 'way/109096915'},
    # https://www.openstreetmap.org/way/28890081 highway=residential and sidewalk=no and maxspeed=30
    # https://www.openstreetmap.org/way/109096915 highway=residential and sidewalk=no and maxspeed=20
    PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_HIGH_SPEED: {'way/25340617', 'way/258562284', 'way/152645928'},
    # https://www.openstreetmap.org/way/25340617 highway=residential and sidewalk=no and maxspeed=50
    # https://www.openstreetmap.org/way/258562284 highway=tertiary and sidewalk=no and maxspeed=50
    # https://www.openstreetmap.org/way/152645928 highway=residential and sidewalk!=* and maxspeed!=*
    PathCategory.INACCESSIBLE: {'way/400711541', 'way/24635973', 'way/25238623', 'way/225895739'},
    # https://www.openstreetmap.org/way/400711541 sidewalk=no and maxspeed:backward=70
    # https://www.openstreetmap.org/way/24635973 foot=no
    # https://www.openstreetmap.org/way/25238623 access=private
    # https://www.openstreetmap.org/way/225895739 service=yes and bus=yes
}


@pytest.fixture(scope='module')
def bpolys():
    """Small bounding boxes."""
    bpolys = gpd.GeoSeries(
        data=[
            shapely.box(8.670396, 49.415353, 8.678818, 49.419778),
            shapely.box(8.693997, 49.414259, 8.697602, 49.415491),
            shapely.box(8.673989, 49.394600, 8.676486, 49.395344),
            shapely.box(8.665670, 49.418945, 8.667255, 49.419226),
            shapely.box(8.695955, 49.394446, 8.703445, 49.396585),
            shapely.box(8.695094, 49.406642, 8.699428, 49.408387),
            shapely.box(8.687238, 49.410277, 8.690776, 49.411319),
            shapely.box(8.6284769976, 49.4267943445, 8.6344636881, 49.4296553394),
            shapely.box(8.693717586876431, 49.41201247400707, 8.70768102673378, 49.408582680835906),
            shapely.box(8.690887, 49.406495, 8.694642, 49.409326),
            shapely.box(8.671805, 49.402299, 8.677725, 49.404730),
            shapely.box(3.354680, 6.444900, 3.369928, 6.455165),
        ],
        crs='EPSG:4326',
    )
    # NOTE: used to generate geojson.io link
    base_url = 'http://geojson.io/#data=data:application/json,'
    encoded = urllib.parse.quote(bpolys.to_json())
    url = base_url + encoded
    logging.debug(url)
    return bpolys


@pytest.fixture(scope='module')
def request_ohsome(bpolys):
    return partial(
        OhsomeClient(
            user_agent='HeiGIT Climate Action Walkability Tester', retry=Retry(total=1)
        ).elements.geometry.post,
        bpolys=bpolys,
        properties='tags',
        time='2024-01-01',
        timeout=60,
    )


@pytest.fixture(scope='module')
def id_filter() -> str:
    """Optimization to make the ohsome API request time faster."""
    full_ids = {''}
    for ids in validation_objects.values():
        full_ids.update(ids)
    full_ids.remove('')
    return f'id:({",".join(full_ids)})'


@pytest.fixture(scope='module')
def osm_return_data(request_ohsome, id_filter) -> gpd.GeoDataFrame:
    ohsome_filter = str(
        '(geometry:line or geometry:polygon) and '
        '(highway=* or route=ferry or railway=platform) and not '
        '(footway=separate or sidewalk=separate or sidewalk:both=separate or '
        '(sidewalk:right=separate and sidewalk:left=separate) or '
        '(sidewalk:right=separate and sidewalk:left=no) or (sidewalk:right=no and sidewalk:left=separate))'
    )
    osm_data = request_ohsome(filter=f'({ohsome_filter})  and ({id_filter})').as_dataframe(multi_index=False)
    osm_data['category'] = osm_data.apply(apply_path_category_filters, axis=1, filters=construct_filters().items())
    return osm_data


@pytest.mark.parametrize('category', validation_objects)
def test_construct_filter_validate(osm_return_data, category):
    osm_return_data = osm_return_data.query('category == @category')

    assert set(osm_return_data['@osmId']) == validation_objects[category]


def test_fetch_osm_data(expected_compute_input, responses_mock):
    with open('resources/test/ohsome_line_response.geojson', 'rb') as vector:
        responses_mock.post(
            'https://api.ohsome.org/v1/elements/geometry',
            body=vector.read(),
        )

    expected_osm_data = gpd.GeoDataFrame(
        data={
            '@other_tags': [{}],
        },
        geometry=[shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)])],
        crs=4326,
    )
    computed_osm_data = fetch_osm_data(expected_compute_input.get_aoi_geom(), 'dummy=yes', OhsomeClient())
    geopandas.testing.assert_geodataframe_equal(computed_osm_data, expected_osm_data, check_like=True)


def test_boost_route_members(expected_compute_input, responses_mock):
    with open('resources/test/ohsome_line_response.geojson', 'rb') as vector:
        responses_mock.post(
            'https://api.ohsome.org/v1/elements/geometry',
            body=vector.read(),
        )

    expected_output = pd.Series(
        data=[
            PathCategory.DEDICATED_EXCLUSIVE,
            PathCategory.DEDICATED_SEPARATED,
            PathCategory.SHARED_WITH_BIKES,
            PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_LOW_SPEED,
            PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_MEDIUM_SPEED,
            PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_HIGH_SPEED,
            PathCategory.INACCESSIBLE,
            PathCategory.MISSING_DATA,
            PathCategory.SHARED_WITH_BIKES,
            PathCategory.SHARED_WITH_BIKES,
        ]
    )

    paths_input = gpd.GeoDataFrame(
        data={
            'category': [
                PathCategory.DEDICATED_EXCLUSIVE,
                PathCategory.DEDICATED_SEPARATED,
                PathCategory.SHARED_WITH_BIKES,
                PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_LOW_SPEED,
                PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_MEDIUM_SPEED,
                PathCategory.SHARED_WITH_MOTORIZED_TRAFFIC_HIGH_SPEED,
                PathCategory.INACCESSIBLE,
                PathCategory.MISSING_DATA,
                PathCategory.SHARED_WITH_BIKES,
                PathCategory.MISSING_DATA,
            ]
        },
        geometry=[
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(0, 0), (1, 0), (1, 1)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
        ],
        crs=4326,
    )
    computed_output = boost_route_members(expected_compute_input.get_aoi_geom(), paths_input, OhsomeClient())
    pd.testing.assert_series_equal(computed_output, expected_output)


def test_boost_route_members_overlapping_routes(expected_compute_input, responses_mock):
    with open('resources/test/ohsome_route_response.geojson', 'rb') as vector:
        responses_mock.post(
            'https://api.ohsome.org/v1/elements/geometry',
            body=vector.read(),
        )

    expected_output = pd.Series(data=[PathCategory.SHARED_WITH_BIKES])

    paths_input = gpd.GeoDataFrame(
        data={'category': [PathCategory.MISSING_DATA]},
        geometry=[
            shapely.LineString([(0, 0), (1, 1)]),
        ],
        crs=4326,
    )
    computed_output = boost_route_members(expected_compute_input.get_aoi_geom(), paths_input, OhsomeClient())
    pd.testing.assert_series_equal(computed_output, expected_output)


def test_fix_geometry_collection():
    expected_geom = shapely.LineString([(0, 0), (1, 0), (1, 1)])

    geometry_collection_input = shapely.GeometryCollection(
        [
            shapely.Point(-1, -1),
            expected_geom,
        ]
    )
    point_input = shapely.Point(-1, -1)

    input_output_map = {
        'unchanged': {'input': expected_geom, 'output': expected_geom},
        'extracted': {'input': geometry_collection_input, 'output': expected_geom},
        'ignored': {'input': point_input, 'output': shapely.LineString()},
    }
    for _, map in input_output_map.items():
        computed_geom = fix_geometry_collection(map['input'])
        assert_geometries_equal(computed_geom, map['output'])


def test_get_color():
    expected_output = pd.Series([Color('#006837'), Color('#feffbe'), Color('#a50026')])

    expected_input = pd.Series([1.0, 0.5, 0.0])
    computed_output = get_color(expected_input)

    pd.testing.assert_series_equal(computed_output, expected_output)


def test_pavement_quality_info_generator():
    verify(generate_detailed_pavement_quality_mapping_info())
