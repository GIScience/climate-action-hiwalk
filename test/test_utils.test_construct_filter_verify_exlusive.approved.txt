(
    highway in (pedestrian,steps,corridor) or
    (
        highway=footway and
        (
    (
        foot in (designated,official) or
        foot!=*
    ) or
    (
        footway in (access_aisle,alley,residential,link,path) or
        footway!=*
    )
    )
    ) or
    (
        highway=path and
        (
    (
        foot in (designated,official) or
        foot!=*
    ) or
    (
        footway in (access_aisle,alley,residential,link,path) or
        footway!=*
    )
    )
    )
    ) and not
    (
    footway in (separate,no) or
    sidewalk=separate or
    sidewalk:left=separate or
    sidewalk:right=separate or
    sidewalk:both=separate or
    access in (no,private,permit,military,delivery,customers) or
    foot in (no,private,use_sidepath,discouraged,destination)
    )
